### ISSUE TITLE
writeFragment within a mutation's update function does not return a ref per documentation

### ISSUE THREAD

**Issue by rmtobin** [apollographql/apollo-client#7099]
I have a set of related items from a query like so:
```
book {
  id
  ...
  related_entity {
    id
    ...
  }
}
```
where in the cache the related_entity field on book is a ref to an EntityNode object. This works and the cache updates as expected, except in the case where on the initial query the `related_entity` field is null and no EntityNode object exists yet, and later through a mutation I create a new EntityNode object and wish to relate it back to the correct BookNode object.

Following the docs (https://www.apollographql.com/docs/react/caching/cache-interaction/#example-adding-an-item-to-a-list) I ended up with code like this:
```
const [mutateEntity] = useMutation(CREATE_OR_UPDATE_ENTITY,{
    update(cache, {data}) {
        cache.modify({
          id: `BookNode:${bookId}`,
          fields: {
            relatedEntity(_i) {
              const newEntityRef = cache.writeFragment({
                fragment: gql`
                  fragment NewEntity on EntityNode {
                    id
                    ...someOtherAttr
                  }`,
                data: data.entityData
              });
              return newEntityRef;
            }
          }
        })
    }
  });
```
but no matter what I seem to try, newEntityRef is always undefined, even though the new EntityNode is definitely in the cache and can be read just fine using the exact same fragment.

I did also find this forum thread from someone with the same issue which led me to believe this may be a regression: https://spectrum.chat/apollo/apollo-client/writefragment-does-not-return-a-reference~2e8cfca9-6067-4136-a60a-13d81264e87a

**Intended outcome:**
Per the docs, calling `cache.writeFragment` within a mutation's update function using the data returned from said mutation should return a reference to the newly written object in the cache (that can then be used to set fields on other cache objects).

**Actual outcome:**
Calling `cache.writeFragment` in the previously mentioned context always returns `undefined`, even though the object is present in the cache and can otherwise be read using `readFragment` or `readQuery`.

**How to reproduce the issue:**
- Create a mutation that creates or updates a cache object
- Within the mutation's update function, attempt to retrieve a ref for the newly created nested object using `writeFragment`.
- Observe that `writeFragment` always returns undefined.

**Versions**
 System:
    OS: macOS 10.15.7
  Binaries:
    Node: 14.3.0 - /usr/local/bin/node
    npm: 6.14.8 - /usr/local/bin/npm
  Browsers:
    Chrome: 85.0.4183.121
    Safari: 14.0
  npmPackages:
    @apollo/client: ^3.2.1 => 3.2.1
    apollo-cache-persist: ^0.1.1 => 0.1.1
    apollo-link-token-refresh: ^0.3.2 => 0.3.2

Prior to updating today I also experienced this in v3.1.1

**Comment by IsaaX** [comment]
@rmtobin Seems like the issue is with apollo-cache-persist, I literally stumbled on this same exact issue today and saw that I was using similar npm packages. I have tests that have apollo-cache-persist disabled which worked but my live environment didn't. After some investigation, I mitigated this by changing from `apollo-cache-persist` to `apollo3-cache-persist`. Seems to have the proper updates to get things back to working.

**Comment by rmtobin** [comment]
@IsaaX You're totally right, I removed `apollo-cache-persist` and this now works as expected. Thanks!  🎉  I didn't even realize the package name had changed for v3 so I didn't realize it needed to be updated when I migrated to v3.

Not sure if this should be closed and/or opened on apollo-cache-persist instead... It is completely non-obvious it is the cache persistor causing undefined values from writeFragment and is probably worth improving the error messaging or the like.
